idseq_pipeline/commands/non_host_alignment_functions.py

```diff
@@ -278,6 +291,80 @@ def log_corrupt(is_corrupt, m8_file, line):
         raise AssertionError
 
 
+def generate_taxon_count_json_from_m8(m8_file, hit_level_file, e_value_type, count_type, stats, output_file):
+    lineage_path = fetch_reference(LINEAGE_SHELF)
+    lineage_map = shelve.open(lineage_path)
+    taxid_properties = {}
+    hit_f = open(hit_level_file, 'rb')
+    m8_f = open(m8_file, 'rb')
+    # lines in m8_file and hit_level_file correspond (same read_id)
+    hit_line = hit_f.readline()
+    m8_line = m8_f.readline()
+    while hit_line and m8_line:
+        hit_line_columns = hit_line.rstrip("\n").split("\t")
+        read_id = hit_line_columns[0]
+        hit_level = hit_line_columns[1]
+        hit_taxid = hit_line_columns[2]
+        if int(hit_level) < 0:
+            hit_line = hit_f.readline()
+            m8_line = m8_f.readline()
+            continue
+        m8_line_columns = m8_line.split("\t")
+        assert m8_line_columns[0] == hit_line_columns[0], "read_ids in %s and %s do not match: %s vs. %s" % (os.path.basename(m8_file),
+                                                            os.path.basename(hit_level_file), m8_line_columns[0], hit_line_columns[0])
+        percent_identity = float(m8_line_columns[2])
+        alignment_length = float(m8_line_columns[3])
+        e_value = float(m8_line_columns[10])
+        if e_value_type != 'log10':
+            e_value = math.log10(e_value)
+        hit_taxid_and_parents = lineage_map.get(hit_taxid, ("-100", "-200", "-300"))[(int(hit_level)-1):3]
+        for i, hit_taxid in enumerate(hit_taxid_and_parents):
+            # TO DO: even if hit is only genus-level, still add a species-level entry for consistency.
+            # The species taxid should be negative to indicate it's not real.
+            # Challenge is to prevent duplicate entries (should be aggregated).
+            # Perhaps a better overall approach would be to record all the hits as before,
+            # but add a flag column to indicate whether the assignment at this level is actually to be trusted
+            # or not (e.g. "mapped to this species ID, but we have determined that the species-level assignment
+            # is meaningless"). Then consume the flag column in the web app to hide/aggregate the meaningless entries.
```

Comment:
i like the idea of assigning it a pseudo speciesid = 1e9 + abs(genus_taxid), or 1e12 + abs(family_id) in the even the genus mapping is also bogus, and flipping the sign to negative is imo optional but i see the value 

adding flags can easily get out of hand and unless the species id is autogenerated it is arbitrary and therefore not as desirable as a deterministic result

Reply:
Actually I'm still debating this with myself...

Reply:
After in-person discussion in today's 1pm meeting we agreed to start with the singleton pseudogenera approach.